Freeform modules

infinisil opened the pull request:
###### Motivation for this change
This implement freeform modules, which are like a combination between `types.submodule` and `types.attrsOf`: The options declared with the submodule are type-checked, while the others are combined using a freeform type. This has been a desired feature in multiple occasions now:
- This will simplify NixOS/rfcs#42 drastically, making the process of declaring options corresponding to `settings` values very conventional.
- Type checking for nixpkgs' `config` argument, as previously attempted in https://github.com/NixOS/nixpkgs/pull/57123 and https://github.com/NixOS/nixpkgs/pull/80582, will now be possible in a sane way.

How it looks like (updated for latest version) (evaluate with `nix-instantiate --eval --strict example.nix -A config`)
```nix
with import ./lib; evalModules {
  modules = [{
    options.settings = mkOption {
      default = {};
      type = with types; submodule {
        # All values that don't have an associated option have to be of this type
        freeformType = attrsOf int;

        options.port = mkOption {
          type = port;
          default = 80;
          description = "The port to use.";
        };
      };
    };

    # Works, even though foo isn't declared as an option
    config.settings.foo = 10;

    # error: The option value `settings.port' in `example.nix' is not of type
    # `16 bit unsigned integer; between 0 and 65535 (both inclusive)'.
    #config.settings.port = "not-a-port";

    ## Overrides default value
    config.settings.port = 8080;
  }];
}

```

Ping @roberth @Ma27 @rycee @Profpatsch @oxij

###### TODO

- [x] Write tests
- [x] Potentially separate orthogonal changes into a separate PR
- [x] Write docs
- [x] Check how this affects performance -> The current implementation is slower than it could be when many freeform values are defined, but since the number of them per module evaluation is expected to be rather low, I'll leave it at this for now.

Infinisil commented:
I now separated the orthogonal changes into https://github.com/NixOS/nixpkgs/pull/82751

roberth commented:
I don't think `attrValueType` is necessary if you merge in one go at the root of the submodule and then `recursiveUpdate` the unstructured values back into `config`.
This puts less burden on the types and, if we don't hardcode `recursiveUpdate`, it allows the module author to keep the unstructured attributes separate if desired.

Infinisil commented:
@roberth I don't think this would work without big changes, because currently it's not known upfront which values are unstructured and which aren't, and this could get complicated with nested `submodules` and `attrsOf` and such.

roberth commented:
@Infinisil `mergeModules` could return the option values and unstructured attrs separately. Doing so should also fix the [laziness problem](https://github.com/NixOS/nixpkgs/pull/82743/files#diff-490491facef0aa6a6dbe8ffc5bd97097R307) caused by generating fake option names for the unstructured definitions.

Quote for when the commit is lost:

> TODO: This leads to infinite recursion in NixOS if it's always turned on, so some use cases won't be supported anymore because of unstructured attrs. Figure out what they are.

One feature I suspect may be lost due to this change is the ability to check for the existence of option declarations before defining them, which I think will be important when modules move into flakes and need to adapt a bit at runtime to their now variable environment.
I've used this to optionally define NixOps deployment keys in a NixOS-compatible module.
I'll make sure we have a test for this laziness property.

Infinisil commented:
@roberth I think I see what you mean, this might be possible, I'll give it a try soon. Implementation would be a bit like: Do a recursive `mergeModules'` for declared options first. After that, determine which definitions haven't been matched with a declaration, and merge those with a single unstructured type. Then combine the results of the `mergeModules'` and the unstructured merge with something like `recursiveUpdate` (it will need to insert the fake option declarations). I don't think it needs to be configurable though, as it will essentially only allow unstructured values under option sets, which makes a lot of sense, and is also the behavior of the current implementation.

infinisil committed:
lib/modules: Internally collect all unmatched definitions

This fundamentally changes how the module evaluation internally
handles definitions without an associated option.

Previously the values of these definitions were discarded and only
the names were propagated. This was fine because that's all that's
needed for optionally checking whether all definitions have an
associated option with _module.check.

However with the upcoming change of supporting freeform modules,
we *do* need the values of these.

With this change, the module evaluation cleanly separates definitions
that match an option, and ones that do not.

Infinisil commented:
@roberth Done that! Commits need some cleanup, and other things are still left to do, but now the code is looking much nicer. I also did some very sweet cleanups related to this. Also I'm now not generating fake options anymore: Only `.config` will now contain unstructured values, `.options` doesn't, which should then also not get in the way of what you mentioned in https://github.com/NixOS/nixpkgs/pull/82802.

Unfortunately without `attrValueType`, a specific error message is significantly worse: With something like
```nix
{
  config._module.unstructuredType = attrsOf int;
  options.a.b = mkOption {};
  config.a.c = "foo";
}
```

it now throws
```
The option value `a' in `example.nix' is not of type `signed integer'.
```

and if you change it to
```nix
{ config.a = 0; }
```
you get (as expected)
```
error: The option path `a' is an attribute set of options,
  but it is defined to not be an attribute set in `example.nix'.
  Did you define its value at the correct and complete path?
```

Instead of the previous (code now in https://github.com/Infinisil/nixpkgs/tree/partially-typed-v2-old)
```
The option `a.c' defined in `example.nix' does not exist
  and the unstructured type `signed integer' of `a' does not support attribute sets.
```

I could get back the original error message if I add `attrValueType` again, but I'm not sure if it's worth it. Maybe that's better left for the future.

roberth commented on lib/modules.nix:
```diff
@@ -58,6 +58,18 @@ rec {
+
+          _module.unstructuredType = mkOption {
+            # Disallow merging for now, but could be implemented nicely with a `types.optionType`
+            type = types.nullOr (types.uniq types.attrs);
```
Is this an approximation of the type of types, with type merging disabled? I think we need `types.type` :)

Infinisil replied:
Since this here would be the only use of that type, and possibly nobody needs merging right now, I think it would be fine to leave it at this approximation for now.

roberth commented on lib/modules.nix:
```diff
@@ -291,19 +333,41 @@ rec {
+      # Transforms unmatchedDefnsByName into a list of definitions
+      unmatchedDefns = concatLists (mapAttrsToList (name: defs:
+        map (def: def // {
+          # Set this so we know when the definition first left unmatched territory
```
You could already modify the `value` as you go.

```
value = { ${name} = def.value; };
```

This removes the need to `map` right before `structuredConfig._module.unstructuredType.merge`. You'd just maintain a consistent `def` all the time.

It may also remove the need for `def.prefix`, depending on whether we need it for error messages.

Infinisil replied:
I had it this way initially, without `prefix`, and generating error messages for invalid options by converting `some.invalid.option = ...` to `loc = [ "some" "invalid" "option" ]`. But that doesn't work well with `some.invalid.option = mkDefault ...`, because `loc` would then be e.g. `[ ... "_type" ]`, because `mkDefault` returns an attrset as well, and there's no way for it to know not to recurse into it. I can imagine other cases where something like this happens. So keeping track of the `prefix` doesn't make this a problem.

Also the prefix would be needed to throw a better error message in the case I mentioned above

roberth commented:
> error message

This can probably be improved quite a bit by rethrowing the error with an extra message.
"The module at ${prefix} allows undeclared definitions to be mixed with known options, but a definition did not satisfy the type for undeclared definitions: ${caughtMsg}. This may also be the result of mistyping an otherwise known option."

Infinisil commented:
@roberth Hm it does work with `builtins.tryEval` and a `deepSeq` on the `unstructuredType.merge ...`, but that's not very nice. I'd have to do some changes to how types throw errors to make it work without `tryEval`.

infinisil committed:
lib/modules: Scope module evaluation variables more tightly

This is a purely cosmetic change so it's easier to see dependencies
between variables.

infinisil committed:
lib/modules: Implement freeform modules

For programs that have a lot of (Nix-representable) configuration options,
a simple way to represent this in a NixOS module is to declare an
option of a type like `attrsOf str`, representing a key-value mapping
which then gets generated into a config file. However with such a type,
there's no way to add type checking for only some key values.

On the other end of the spectrum, one can declare a single separate
option for every key value that the program supports, ending up with a module
with potentially 100s of options. This has the benefit that every value
gets type checked, catching mistakes at evaluation time already. However
the disadvantage is that the module becomes big, becomes coupled to the
program version and takes a lot of effort to write and maintain.

Previously there was a middle ground between these two
extremes: Declare an option of a type like `attrsOf str`, but declare
additional separate options for the values you wish to have type
checked, and assign their values to the `attrsOf str` option. While this
works decently, it has the problem of duplicated options, since now both
the additional options and the `attrsOf str` option can be used to set a
key value. This leads to confusion about what should happen if both are
set, which defaults should apply, and more.

Now with this change, a middle ground becomes available that solves above
problems: The module system now supports setting a freeform type, which
gets used for all definitions that don't have an associated option. This
means that you can now declare all options you wish to have type
checked, while for the rest a freeform type like `attrsOf str` can be
used.

infinisil committed:
lib/types: Make submodules use the freeform type description

Submodules that have a freeform type set behave as if that was the type
of the option itself (for values that don't have an option). Since the
submodules options are shown as separate entries in the manual, it makes
sense to show the freeform type as the submodule type.

infinisil committed:
lib/tests: Add tests for freeform modules

Infinisil commented:
I think I'll have to get the `attrValueType` thing back after all. Not only does it allow for much better error messages in some cases, but after doing a performance check, I also found that the whole definition coercion thing is really slow due to the many list operations.

Infinisil commented:
@roberth I think let's not do this `attrValueType` thing after all, for now. The current implementation here works well and has clean and simple code, the only problem is that it can be a bit slow if there are many unstructured attributes, but that will be a rare occurence. If this *does* becomes a problem in the future, it can still be changed then, though ideally this would also bring some module system improvements along with it so `attrValueType` can be implemented nicely (which it currently can't).

So I think the only thing left to do here is to write some docs. Also I think the naming is a bit off, it's not "partially typed modules", because everything is still type checked, there's just a fallback type now for undeclared options. "unstructured attributes" is also a bit unfitting. Maybe the feature could be called "Fallback-typing" instead.

Infinisil commented:
I have now written some initial docs for this feature. Once the docs are finalized and reviewed, I think this is ready to be merged.

Infinisil commented:
![Screen Shot 2020-05-22 at 03 41 25](https://user-images.githubusercontent.com/20525370/82622458-3310b780-9bde-11ea-8825-9db92a9322d8.png)

roberth commented on nixos/doc/manual/development/fallback-typing.xml:
```diff
@@ -0,0 +1,135 @@
+  </programlisting>
+  <note>
+   <para>
+    While the type <link linkend="types-lazyAttrsOf"><literal>types.lazyAttrsOf</literal></link> can be used to prevent infinite recursion as well, this has some downsides to it, and is not recommended unless the consequences are fully understood.
```
```suggestion
    You could use the <link linkend="types-lazyAttrsOf"><literal>types.lazyAttrsOf</literal></link> to prevent infinite recursion, but it has only partial support for conditional definitions with <literal>mkIf</literal>.
```

No need to scare the reader.

roberth reviewed (commented):
The documentation needs a bit of work. I'd summarize this feature with

**$feature allows module users to define values for options that have not been declared explicitly by the module author.**

The new term 'fallback _typing_' does not give me a clue that I need that feature when I want to allow arbitrary _options_. This feature is more about modules than types in my view (besides options of course).

The docs read like you're describing a change, like you would in an rfc or a commit message. Could you write it as though it has always existed? That way it's easier to be straight to the point.
If you start the section in the manual with a brief description of the feature, you'll help the reader quite a lot. You could use a sentence like the highlighted one above.

Infinisil commented:
> The docs read like you're describing a change, like you would in an rfc or a commit message

Yeah that's what I noticed too, will definitely try to improve on that

pasqui23 commented:
Another use would be to use packages as submodules,like what I would like to do in https://github.com/NixOS/nixpkgs/pull/89943

Infinisil commented:
@roberth What do you think would be a better term for this feature? Maybe one of:
- A fallback option type
- Declarationless definitions
- Optionless definitions
- Freeform definitions
- Freeform configuration
- Freeform modules

The last one sounds pretty good imo

roberth commented:
> Freeform modules
The last one sounds pretty good imo

:100:

Infinisil commented:
I'll wait until https://github.com/NixOS/nixpkgs/pull/75584 is merged so I can extend the "Options for Program Settings" section with this feature.

infinisil committed:
lib/modules: Fix freeform modules when there's no definitions

infinisil committed:
nixos/doc: Add freeform modules documentation

Co-Authored-By: Robert Hensing <robert@roberthensing.nl>

Infinisil commented:
Now updated to term this feature freeform modules instead and revised the documentation to be less RFC-y. Also fixed a bug with freeform modules. Once the documentation is reviewed, I think this PR is ready for merging.

New section under "Writing modules":
![Screen Shot 2020-08-04 at 01 39 38](https://user-images.githubusercontent.com/20525370/89237441-6ff02600-d5f3-11ea-94b0-4d2d3f585f10.png)

New section under the `settings` option docs (introduced in #75584):
![Screen Shot 2020-08-04 at 01 39 56](https://user-images.githubusercontent.com/20525370/89237484-92823f00-d5f3-11ea-92c8-e87155752eca.png)

roberth commented on nixos/doc/manual/development/freeform-modules.xml:
```diff
@@ -0,0 +1,69 @@
+  Freeform modules allow you to define values for option paths that have not been declared explicitly. This can be used to add attribute-specific types to <literal>attrsOf</literal> options.
+ </para>
+ <para>
+  This feature can be enabled by using the option <literal>_module.freeformType</literal> to define a freeform type. By doing this, all assigments without an associated option will get merged using the freeform type and combined into the resulting <literal>config</literal> set. Since this feature influences the whole module evaluation, it is mostly recommended for submodule options only.
```
```suggestion
  This feature can be enabled by using the option <literal>_module.freeformType</literal> to define a freeform type. By doing this, all assigments without an associated option will be merged using the freeform type and combined into the resulting <literal>config</literal> set. Since this feature nullifies name checking for entire option trees, it is only recommended for use in submodules.
```

roberth commented on nixos/doc/manual/development/freeform-modules.xml:
```diff
@@ -0,0 +1,69 @@
+ <example xml:id="ex-freeform-module">
+  <title>Freeform submodule</title>
+  <para>
+   The following shows a submodule assigning a freeform type that allows arbitrary <literal>str</literal> attribute values, but also declares an option for the <literal>port</literal> attribute to have it type-checked and assign a default value.
```
```suggestion
   The following shows a submodule assigning a freeform type that allows arbitrary attributes with <literal>str</literal> values below <literal>settings</literal>, but also declares an option for the <literal>settings.port</literal> attribute to have it type-checked and assign a default value.
```

roberth commented on nixos/doc/manual/development/freeform-modules.xml:
```diff
@@ -0,0 +1,69 @@
+    default = {};
+    type = lib.types.submodule {
+
+      config._module.freeformType = with lib.types; attrsOf str;
```
```suggestion
      freeformType = with lib.types; attrsOf str;
```

Can we add syntax sugar like this? Besides being a bit cryptic, it's kind of in the wrong part of the module syntax.

(This suggestion applies to other mentions of `_module.freeformType` as well of course)

Infinisil replied:
I haven't considered this yet. I don't see an immediate problem with it, other than it making the implementation a bit more involved. I'll give it some thought

roberth commented on nixos/doc/manual/development/settings-options.xml:
```diff
@@ -167,13 +167,51 @@ in {
+   <section xml:id="sec-settings-attrs-options">
+    <title>Option declarations for attributes</title>
+    <para>
+     It can be very valuable to declare options for a subset of <literal>settings</literal> attributes in order to check their type and show their documentation in the manual. This can be done using <xref linkend='sec-freeform-modules'/>.
```
```suggestion
     Some <literal>settings</literal> attributes may deserve some extra care. They may need a different type, default or merging behavior, or they are essential options that should show their documentation in the manual. This can be done using <xref linkend='sec-freeform-modules'/>.
```

roberth commented on nixos/doc/manual/development/settings-options.xml:
```diff
@@ -167,13 +167,51 @@ in {
+      </para>
+<programlisting>
+settings = lib.mkOption {
+  # Setting this type allows for correct merging behavior
```
```suggestion
```

Although defaults can be considered _merging behavior_ it's not so evident from the example. I've mentioned merging behavior in an earlier suggestion, so that may be sufficient to cover it. Though perhaps a different example can demonstrate the value of custom merging behavior?

roberth reviewed (commented):
It's almost there, just
 - docs suggestions
 - syntax sugar `freeformType` <-> `config._module.freeformType`

infinisil committed:
lib/modules: Add syntactic sugar for config._module.freeformType

This introduces `freeformType` as a top-level module attribute, allowing
definitions like

  {
    freeformType = ...;
    options = ...;
    config = ...;
  }

Infinisil commented:
@roberth Thanks for all the suggestions, I implemented them now. After some thought, I think there's no reason not to introduce some syntactic sugar for `freeformType`, so I did that now. Was fairly easy to add even.

roberth commented:
The freeform type doesn't seem to be visible in the generated option docs. Could we generate a line for an "option" like `"${somePrefix}.*"` there?
I think we've discussed this in an earlier pr, but I couldn't find it.

We may want to use a freeform "option" rather than a freeform type for this reason, because that provides an opportunity to provide the description string and an example.

Infinisil commented:
@roberth With https://github.com/NixOS/nixpkgs/pull/82743/commits/2d45a62899d47c109a0b8ce4ca9d33265b8a1a37 the freeform type is shown as the submodules type. E.g. with a config like
```nix
{ lib, ... }: {
  options.fofofofo = lib.mkOption {
    type = lib.types.submodule {
      freeformType = lib.types.attrsOf lib.types.str;
      options.babababa = lib.mkOption {};
    };
  };
  config = {
    documentation.nixos.includeAllModules = true;
    fileSystems."/".device = "ntest";
    boot.loader.grub.device = "nodev";
  };
}
```

Building with `nix-build '<nixpkgs/nixos>' --arg configuration ./config.nix -A config.system.build.manual.manualHTML` gives

![2020-08-12_20-27](https://user-images.githubusercontent.com/20525370/90053100-4eb9c480-dcda-11ea-9a8e-5893c6c7d86f.png)

roberth commented on nixos/doc/manual/development/freeform-modules.xml:
```diff
@@ -0,0 +1,69 @@
+{ lib, config, ... }: {
+
+  options.settings = lib.mkOption {
+    default = {};
```
This example will be used as a starting point by some module authors. Ideally it'd be a complete example like the one in settings-options.
Adding a description and example here makes the example more complete, but less relevant to the text.
Perhaps instead of setting the default, we could use that line to cross-reference the more complete example `ex-settings-typed-attrs` to use as a starting point.

Infinisil replied:
I added a reference to that example.

roberth reviewed (commented):
Seems like this can and should be handled by the outer `mkOption`. I was focusing on the `submodule` invocation whereas the generated option line stems from the option that contains it.

Maybe I'm not the only one to make that mistake so I've put some suggestions to improve the docs to help module authors write docs.
I noticed that we'll have some repetition in the descriptions, so `formats` can be helpful by providing standard doc strings that can be interpolated into those.
It shouldn't be too much of a rabbithole to add that to `formats`, but it's also ok to just make the description in the example more complete.

Infinisil commented on nixos/doc/manual/development/settings-options.xml:
```diff
@@ -167,13 +167,50 @@ in {
+  description = ''
+    Configuration for foo, see
+    &lt;link xlink:href="https://example.com/docs/foo"/&gt;
+    for supported values.
```
I see what you mean, but I think this could be done as a separate PR.

roberth reviewed (approved):
Great work @Infinisil!

The technical aspects are in order and judging from https://github.com/NixOS/rfcs/pull/42#issuecomment-665110193, this feature addition is not controversial, so a merge seems more than appropriate. :tada:

Infinisil commented:
Awesome! Thanks for all the reviewing and merging @roberth!

marsam commented:
This broke nix-darwin, bisected to https://github.com/NixOS/nixpkgs/commit/fd75dc876586bde8cdb683a6952a41132e8db166

```
$ darwin-rebuild build --dry-run
building the system configuration...
error: The option `nixpkgs.localSystem' defined in `<unknown-file>' does not exist.
(use '--show-trace' to show detailed location information)
```
<details>
<summary>full trace</summary>

```
$ darwin-rebuild build --dry-run --show-trace
building the system configuration...
error: while evaluating the attribute 'activationScript' of the derivation 'darwin-system-20.09.git.78922ca+darwin3.47fe6db' at /Users/marsam/code/nix-config/darwin/modules/system/default.nix:74:7:
while evaluating the attribute 'activationScripts.script.text' at undefined position:
while evaluating 'g' at /Users/marsam/code/nix-config/nixpkgs/lib/attrsets.nix:276:19, called from undefined position:
while evaluating anonymous function at /Users/marsam/code/nix-config/nixpkgs/lib/modules.nix:98:72, called from /Users/marsam/code/nix-config/nixpkgs/lib/attrsets.nix:279:20:
while evaluating the attribute 'value' at /Users/marsam/code/nix-config/nixpkgs/lib/modules.nix:450:9:
while evaluating the option `system.activationScripts.script.text':
while evaluating the attribute 'mergedValue' at /Users/marsam/code/nix-config/nixpkgs/lib/modules.nix:482:5:
while evaluating the attribute 'values' at /Users/marsam/code/nix-config/nixpkgs/lib/modules.nix:476:9:
while evaluating the attribute 'values' at /Users/marsam/code/nix-config/nixpkgs/lib/modules.nix:575:7:
while evaluating anonymous function at /Users/marsam/code/nix-config/nixpkgs/lib/modules.nix:462:28, called from /Users/marsam/code/nix-config/nixpkgs/lib/modules.nix:462:17:
while evaluating definitions from `/Users/marsam/code/nix-config/darwin/modules/system/activation-scripts.nix':
while evaluating 'dischargeProperties' at /Users/marsam/code/nix-config/nixpkgs/lib/modules.nix:534:25, called from /Users/marsam/code/nix-config/nixpkgs/lib/modules.nix:463:137:
while evaluating the attribute 'value' at /Users/marsam/code/nix-config/nixpkgs/lib/modules.nix:327:44:
while evaluating the attribute 'activationScripts.applications.text' at undefined position:
while evaluating 'g' at /Users/marsam/code/nix-config/nixpkgs/lib/attrsets.nix:276:19, called from undefined position:
while evaluating anonymous function at /Users/marsam/code/nix-config/nixpkgs/lib/modules.nix:98:72, called from /Users/marsam/code/nix-config/nixpkgs/lib/attrsets.nix:279:20:
while evaluating the attribute 'value' at /Users/marsam/code/nix-config/nixpkgs/lib/modules.nix:450:9:
while evaluating the option `system.activationScripts.applications.text':
while evaluating the attribute 'mergedValue' at /Users/marsam/code/nix-config/nixpkgs/lib/modules.nix:482:5:
while evaluating the attribute 'values' at /Users/marsam/code/nix-config/nixpkgs/lib/modules.nix:476:9:
while evaluating the attribute 'values' at /Users/marsam/code/nix-config/nixpkgs/lib/modules.nix:575:7:
while evaluating anonymous function at /Users/marsam/code/nix-config/nixpkgs/lib/modules.nix:462:28, called from /Users/marsam/code/nix-config/nixpkgs/lib/modules.nix:462:17:
while evaluating definitions from `/Users/marsam/code/nix-config/darwin/modules/system/applications.nix':
while evaluating 'dischargeProperties' at /Users/marsam/code/nix-config/nixpkgs/lib/modules.nix:534:25, called from /Users/marsam/code/nix-config/nixpkgs/lib/modules.nix:463:137:
while evaluating the attribute 'value' at /Users/marsam/code/nix-config/nixpkgs/lib/modules.nix:327:44:
while evaluating the attribute 'passAsFile' of the derivation 'system-applications' at /Users/marsam/code/nix-config/nixpkgs/pkgs/build-support/trivial-builders.nix:7:7:
while evaluating the attribute 'buildCommand' of the derivation 'darwin-manpages' at /Users/marsam/code/nix-config/nixpkgs/pkgs/build-support/trivial-builders.nix:7:7:
while evaluating the attribute 'buildCommand' of the derivation 'manual-olinkdb' at /Users/marsam/code/nix-config/nixpkgs/pkgs/build-support/trivial-builders.nix:7:7:
while evaluating the attribute 'buildCommand' of the derivation 'darwin-manual-combined' at /Users/marsam/code/nix-config/nixpkgs/pkgs/build-support/trivial-builders.nix:7:7:
while evaluating the attribute 'buildCommand' of the derivation 'generated-docbook' at /Users/marsam/code/nix-config/nixpkgs/pkgs/build-support/trivial-builders.nix:7:7:
while evaluating the attribute 'buildCommand' of the derivation 'options-db.xml' at /Users/marsam/code/nix-config/nixpkgs/pkgs/build-support/trivial-builders.nix:7:7:
while evaluating 'flip' at /Users/marsam/code/nix-config/nixpkgs/lib/trivial.nix:138:16, called from /Users/marsam/code/nix-config/darwin/doc/manual/default.nix:48:21:
while evaluating 'optionAttrSetToDocList'' at /Users/marsam/code/nix-config/nixpkgs/lib/options.nix:140:37, called from /Users/marsam/code/nix-config/darwin/doc/manual/default.nix:9:72:
while evaluating 'collect' at /Users/marsam/code/nix-config/nixpkgs/lib/attrsets.nix:178:19, called from /Users/marsam/code/nix-config/nixpkgs/lib/options.nix:162:67:
while evaluating 'isType' at /Users/marsam/code/nix-config/nixpkgs/lib/types.nix:14:18, called from /Users/marsam/code/nix-config/nixpkgs/lib/attrsets.nix:179:8:
while evaluating 'evalModules' at /Users/marsam/code/nix-config/nixpkgs/lib/modules.nix:21:17, called from /Users/marsam/code/nix-config/darwin/modules/documentation/default.nix:19:24:
The option `nixpkgs.localSystem' defined in `<unknown-file>' does not exist.
```
</details>

The only place nix-darwin uses `nixpkgs.localSystem` is in the manual https://github.com/LnL7/nix-darwin/blob/47fe6db9c9cd22c4ba57df3281f98f454880bea7/modules/documentation/default.nix#L20. 

I'm not familiar with modules internals, so I'm appreciate any hint to for a proper fix

cc: @LnL7

roberth commented:
@marsam I can't find a declaration for `localSystem` in nix-darwin, so the error message seems to be correct. Grepping only reveals the usage site you linked but no declaration.
The file analogous to Nixpkgs where I would expect it to be defined is https://github.com/LnL7/nix-darwin/blob/master/modules/nix/nixpkgs.nix
I don't think nix-darwin needs that option, so I've created a PR for the simple solution https://github.com/LnL7/nix-darwin/pull/216

NixOS-specific rant:
If a module needs access to cross-compilation info, it should generally be taken from `pkgs` directly; either `pkgs.stdenv.hostPlatform`, which is the what the eventual machine runs or if you really do need to care about where the config is built: `pkgs.stdenv.buildPlatform`. Generally you care about `pkgs.stdenv.hostPlatform` which can not be assumed to be equivalent to `config.nixpkgs.localSystem` because of `config.nixpkgs.crossSystem`. `local` and `cross` only make sense in the context of `nixos-rebuild --target-host`, which uses incompatible terminology...

vcunat commented:
TL;DR: I assume we'll revert this merge until the issues get resolved?

This merge also broke Hydra's evaluation of trunk-combined.  It's some kind of interaction with other conditions and I haven't managed to reproduce it locally so far though @samueldr did (maybe this just triggered some Nix bug)... but the merge commit of this PR (6d0a85fe52) won't evaluate on Hydra whereas the parent commit [succeeded](https://hydra.nixos.org/eval/1606980).

The message is
```
error: [json.exception.type_error.302] type must be string, but is null
```
and it seems likely that it comes from evaluating `nixos/release-combined.nix -A nixos.manual.x86_64-linux`.  Some context was [on IRC yesterday](https://logs.nix.samueldr.com/nixos-dev/2020-08-15#3866497;).

Infinisil commented:
@roberth Nice detective work! Both the nix-darwin and the hydra failures seem to originate from https://github.com/NixOS/nixpkgs/commit/fd75dc876586bde8cdb683a6952a41132e8db166, which apparently checks options stricter, but I'm not quite sure yet why (I'll investigate this a bit more)

Fyi, the hydra failure originates from this command now failing:
```
$ nix-instantiate nixos/tests/boot.nix -A biosCdrom
error: The option `virtualisation.qemu.consoles' defined in `/home/infinisil/src/nixpkgs/nixos/modules/testing/test-instrumentation.nix' does not exist.
```

Which is indeed fixed by @roberth's PR. Similar for nix-darwin.

roberth commented:
> TL;DR: I assume we'll revert this merge until the issues get resolved?

Shouldn't be necessary.

This change is merely exposing undeclared options that weren't discovered by the previous implementation.
I've merged a fix for the one that was blocking release-small. `trunk-combined` should also be fixed now, because it doesn't seem to suffer from other errors of the same time. We'll see when evaluation is done.

 (I'm wondering, shouldn't TL;DR be a summary?)

Infinisil commented:
Ah figured it out: The previous implementation only checked option existence for config paths that are evaluated, while the new one checks all of them unconditionally. Here's an example that illustrates this:
```nix
(import <nixpkgs/lib>).evalModules {
  modules = [({ lib, ... }: {
    options.root = lib.mkOption {};
    options.foo.bar = lib.mkOption {};

    config.root = "root";
    config.foo.qux = "qux";
  })];
}
```

Evaluating with `nix-instantiate --eval -A config.root` previously worked because `config.foo` wasn't evaluated, so it would still fail with `nix-instantiate --eval -A config.foo`. With the new implementation, both commands fail. I think this change in behavior is justified, because there can be other reasons an option doesn't exist other than the config path not being evaluated (e.g. misspellings). In fact I think this change might even fix https://github.com/NixOS/nixpkgs/issues/36590.

It looks like previously this was (ab)used to make evaluation not fail: https://github.com/NixOS/nixpkgs/blob/aeffd67cec5b76e327f244c7464a4254113836cc/nixos/modules/testing/test-instrumentation.nix#L11-L18

vcunat commented:
Thanks, that was really fast.

> (I'm wondering, shouldn't TL;DR be a summary?)

I did consider that as something like summary/conclusion, explaining the details below, though I agree I typically see it used in a bit different way.

roberth commented:
@Infinisil I think @LnL7 has a good point here https://github.com/LnL7/nix-darwin/pull/216#issuecomment-674966068 that this is an api change. Even if the old api is undocumented and untested, it may have an important role in modular code bases. Do you think it's feasible to warn about the old behavior or to emulate it?
If only warning about it is feasible, we can recommend to extract the option into a separate file, which [may](https://en.wikipedia.org/wiki/Dependency_inversion_principle) be a good idea.

Infinisil commented:
I don't think trying to emulate the old behavior is a good idea, because from what we've seen already, that behavior was very prone to mistakes (https://github.com/NixOS/nixpkgs/issues/36590 and https://github.com/LnL7/nix-darwin/pull/216). While https://github.com/NixOS/nixpkgs/pull/95584 wasn't a mistake per se, it still relied on a hack to make it work.

I think it would be a good idea to mention this change in behavior though, probably in the breaking changes section in the release notes.

dali99 commented:
There doesn't seem to be any way to define typing for an optional setting, I was looking to set  `default = unset` in some way, since some configuration handle `null` and "not set" differently.

Infinisil commented:
@dali99 Indeed, that's something that could be improved in the future. See also https://github.com/NixOS/nixpkgs/pull/63553 which is related.
